packages/runtime/container-runtime/src/gc/garbageCollection.ts

```diff
@@ -482,53 +490,172 @@ export class GarbageCollector implements IGarbageCollector {
+
+				/** Post-GC step. Includes logging events and other cleanup. */
+				await this.runPostGCSteps(logger);
 				return gcStats;
 			},
 			{ end: true, cancel: "error" },
 		);
 	}
 
+	/**
+	 * Runs any steps before actual GC starts. It initializes state from base snapshot (required first time GC runs)
+	 * and notifies runtime that GC is starting.
+	 */
 	private async runPreGCSteps() {
 		// Ensure that state has been initialized from the base snapshot data.
 		await this.initializeGCStateFromBaseSnapshotP;
 		// Let the runtime update its pending state before GC runs.
 		await this.runtime.updateStateBeforeGC();
 	}
 
-	private async runPostGCSteps(
-		gcData: IGarbageCollectionData,
-		gcResult: IGCResult,
-		logger: ITelemetryLogger,
+	/**
+	 * Runs garbage collection. It generates the reference graph, runs GC algorithm on it to find references and
+	 * unreferenced nodes and runs Mark and Sweep phases.
+	 */
+	private async runGC(
+		fullGC: boolean,
 		currentReferenceTimestampMs: number,
+		logger: ITelemetryLogger,
 	): Promise<IGCStats> {
-		// Generate statistics from the current run. This is done before updating the current state because it
-		// generates some of its data based on previous state of the system.
+		const gcData = await this.runtime.getGCData(fullGC);
+		const gcResult = runGarbageCollection(gcData.gcNodes, ["/"]);
+
+		// Generate stats before running Mark / Sweep phase because it uses previous GC state which will get updated
+		// in these stages.
 		const gcStats = this.generateStats(gcResult);
 
-		// Update the current mark state and update the runtime of all used routes or ids that used as per the GC run.
-		const sweepReadyNodes = this.updateMarkPhase(
+		// Run the Mark phase. It will mark nodes as referenced / unreferenced appropriately and return a list of
+		// node ids that are ready to be swept.
+		const sweepReadyNodeIds = this.runMarkPhase(
 			gcData,
 			gcResult,
 			currentReferenceTimestampMs,
 			logger,
 		);
+
+		// Run the Sweep phase. It will delete sweep ready nodes and return updated GC data - gcData which has deleted
+		// nodes removed.
+		const updatedGCData = this.runSweepPhase(
+			gcData,
+			gcResult,
+			sweepReadyNodeIds,
+			currentReferenceTimestampMs,
+			logger,
+		);
+
+		this.gcDataFromLastRun = cloneGCData(updatedGCData);
+		return gcStats;
+	}
+
+	/**
+	 * Runs any steps after GC completes. It logs unreferenced events, increments counters and does other cleanup.
+	 */
+	private async runPostGCSteps(logger: ITelemetryLogger) {
+		// Log pending unreferenced events such as a node being used after inactive. This is done after GC runs and
+		// updates its state so that we don't send false positives based on intermediate state. For example, we may get
+		// reference to an unreferenced node from another unreferenced node which means the node wasn't revived.
+		await this.telemetryTracker.logPendingEvents(logger);
+		this.newReferencesSinceLastRun.clear();
```

Comment:
This moved from middle of Mark step to the end. Glancing at the code it seems fine, just curious if it was necessary or just seemed cleaner.

Reply:
It wasn't necessary. It's just cleaner. It needs to reset after GC completes.